src/strategies/latency.js

```diff
@@ -0,0 +1,231 @@
+ * 	logger: true,
+ * 	registry: {
+ * 		strategy: "LatencyStrategy",
+ * 		strategyOptions: {
+ * 			sampleCount: 5,
+ * 			lowLatency: 10,
+ * 			collectCount: 5,
+ * 			pingInterval: 10
+ * 		}
+ * 	}
+ * });
+ *
+ * @class LatencyStrategy
+ */
+class LatencyStrategy extends BaseStrategy {
+
+	constructor(registry, broker) {
+		super(registry, broker);
+
+		this.opts = _.defaultsDeep(registry.opts.strategyOptions, {
+			sampleCount: 5,
+			lowLatency: 10,
+			collectCount: 5,
+			pingInterval: 10
+		});
+
+		this.brokerStopped = false;
+
+		this.hostLatency = new Map();
+
+		/* hostMap contains:
+			hostname => {
+				historicLatency: [],
+				nodeList: []
+			}
+		*/
+		this.hostMap = new Map();
+
+		// short circuit
+		if (!this.broker.transit) return;
+
+		if (this.broker.localBus.listenerCount("$node.latencyMaster") === 0) {
+			// claim as master
+			this.broker.localBus.on("$node.latencyMaster", function() {});
+			// respond to PONG
+			this.broker.localBus.on("$node.pong", this.processPong.bind(this));
+			// dynamically add new node
+			this.broker.localBus.on("$node.connected", this.addNode.bind(this));
+			// dynamically remove node
+			this.broker.localBus.on("$node.disconnected", this.removeHostMap.bind(this));
+			// try to discovery all nodes on start up
+			this.broker.localBus.on("$broker.started", this.discovery.bind(this));
+			// clean up ourselves
+			this.broker.localBus.on("$broker.stopped", () => this.brokerStopped = true);
+		} else {
+			// remove node if we are told by master
+			this.broker.localBus.on("$node.latencySlave.removeHost", this.removeHostLatency.bind(this));
+		}
+
+		this.broker.localBus.on("$node.latencySlave", this.updateLatency.bind(this));
+	}
+
+	// Master
+	discovery() {
+		return this.broker.transit.sendPing().then(function() {
+			setTimeout(this.pingHosts.bind(this), 1000 * this.opts.pingInterval);
+		}.bind(this));
+	}
+
+	// Master
+	pingHosts() {
+
+		if (this.brokerStopped) return;
+		/*
+			Smart Ping: only ping the host, not the nodes (which may be many)
+
+			Although, if that particular node on the host is overloaded,
+			the measurement may be skewed.
+		*/
+		let hosts = [];
+		if (this.hostMap.size > 0) {
+			hosts = this.hostMap.values();
+		}
+
+		this.broker.Promise.map(hosts, function(host) {
+			return this.broker.transit.sendPing(host.nodeList[0]);
+		}.bind(this), { concurrency: 5 }).then(function() {
+			setTimeout(this.pingHosts.bind(this), 1000 * this.opts.pingInterval);
+		}.bind(this));
+	}
+
+	// Master
+	processPong(payload) {
+		let node = this.registry.nodes.get(payload.nodeID);
+		if (!node) return;
+
+		let avgLatency = null;
+
+		this.mapNode(node);
+
+		let hostMap = this.hostMap.get(node.hostname);
+
+		if (hostMap.historicLatency.length > (this.opts.collectCount - 1))
+			hostMap.historicLatency.shift();
+
+		hostMap.historicLatency.push(payload.elapsedTime);
+
+		avgLatency = hostMap.historicLatency.reduce(function(sum, latency) {
```

Comment:
Please use shorthand functions in **all** places without `bind`: `(sum, latency) => {... }`